### ISSUE TITLE
String invoke in nested scope wrongly resolves to parent scope

### ISSUE THREAD

**Issue by eskatos** [gradle/kotlin-dsl#1119]
### Expected Behavior
```kotlin
dependencies {
    constraints {
        "implementation"("com.google.collections:google-collections") {
            version { rejectAll() }
            because("Google collections is superceded by Guava")
        }
    }
}
```

to add a dependency constraints on the `implementation` configuration.

### Current Behavior
It compiles cleanly but adds a dependency instead of a dependency constraint because the string invoke invocation resolves to the one from the outer `DependencyHandlerScope` instead of the one from the inner `DependencyConstraintHandlerScope`.

### Context
See #710

**Comment by efemoney** [comment]
I'm not sure if Kotlin's `DslMarker` annotation will come in handy here. Maybe?

**Comment by bamboo** [comment]
@efemoney Right, good thinking. Currently we cannot simply annotate the required Gradle interfaces (because dependencies) but it's an interesting path to explore, thanks.

**Comment by efemoney** [comment]
Yes thats true. Initially I faced this problem with `NamedDomainObjectContainerScope` and annotating that class would've fixed it but I see how thats not entirely pragmatic considering the actual Gradle interfaces.
